### ISSUE TITLE
Comparison takes long on dates over 2036

### ISSUE THREAD

**Issue by markov2** [houseabsolute/DateTime.pm#90]
When you use a date after 2036, the DateTime::_compare() gets called very often.  This is best visible when the year is very large, as in the program below which makes 100,000 calls.
```perl
#!/usr/bin/perl
  
use strict;
use warnings;

use DateTime;
use Time::HiRes qw(time);

my $dt1 = DateTime->new(year => 9999, month => 1, day => 1);
my $dt2 = DateTime->new(year => 2019, month => 5, day => 28, time_zone => 'Europe/London');

my $start  = time;
my $x = $dt1 < $dt2 ? $dt1 : $dt2;

my $elapse = sprintf "%dms", (time - $start)*1000;
warn "$elapse: $x\n";
```

**Comment by autarch** [comment]
This actually isn't happening in DateTime.pm, it's because of how DateTime::TimeZone works. If the time zone has DST rules, then we need to calculate all the DST transitions until the timestamp for the DT.

This particular case is triggering the issue because when you compare a DT with a floating TZ to one without, it sets the DT with the floating TZ to the same TZ as the DT we're comparing against. Otherwise there's no way to compare the two.

It doesn't really make sense to use a time zone more than a few years in the future, as time zone rules change regularly. It _definitely_ doesn't make sense to use them for dates that are decades away.

There is a warning if you try to create a far future DT with a non-UTC or floating TZ. I think it'd make sense to also warn in other cases that trigger this behavior, like the comparison case here.

**Comment by markov2** [comment]
Do you really need to calculate time-zones when the years differ?

**Comment by autarch** [comment]
> Do you really need to calculate time-zones when the years differ?

Of course. If one TZ is at +12 and the other at -12 then you could easily have a case where the years differ but the one with a lower year is actually later in UTC.

**Comment by arodland** [comment]
But then again, not if the years differ by more than 1. And indeed probably not if the ymd-part differs by 3 or more days :)

**Comment by markov2** [comment]
But can you explain why, in above script <code>$dt1 < $dt2</code> takes 7 seconds on my system, but <code>$dt1->epoch < $dt2->epoch</code> less than one millisecond?

Wouldn't it simplify the code  a lot when <code>_compare()</code> would simple call <code>epoch()</code> ?

**Comment by autarch** [comment]
> But can you explain why, in above script $dt1 < $dt2 takes 7 seconds on my system, but $dt1->epoch < $dt2->epoch less than one millisecond?

I already did. It's because the DT in the floating time zone is converted to the time zone of the other DT, so it's equivalent to calling `$dt1->clone->set_time_zone('Europe/London')->epoch < $dt2->epoch`.

**Comment by abeverley** [comment]
Is there any mileage in doing some sort of "shortcut" before the timezone is added/calculated? E.g. if the day/month/year are off by so much that the timezone is irrelevant. Even something simple like if the years are different by more than 2 would work for our use case. Thanks.

**Comment by markov2** [comment]
Certainly with an uncertain future of DST, wouldn't it be safe to simply ignore their existence for years to come?  At least: until now, all DST effects have neutralized after the end of each year: no nett effect... it is only required to calculate the last year.  Not all intermediate years.

(In our application, people can describe events will happen in the future (years).  We do not want that to be inconveniently slow)

**Comment by autarch** [comment]
@abeverley - yes, that sort of fix could work. I think just checking if the year diff is >1 would be sufficient.

**Comment by autarch** [comment]
@markov2 I'm not sure how one would decide when to start ignoring time zones. This really is something that users of the library are in a much better position to do than me as the library author.
